### PR TITLE
Prompt buffer bug fixes

### LINKED ISSUE #2742: Fix `nyxt/prompt-buffer-mode:{next,previous}-page`
Commands `nyxt/prompt-buffer-mode:next-page` and `nyxt/prompt-buffer-mode:previous-page` don't work as expected.  They don't update the list of suggestions by a full page (i.e. by all visible ones).  Also, the current suggestion should be set to the index of the current suggestion (with respect to the visible suggestions) before running the command.

The JS logic in `step-page-index` needs to be reviewed.

### LINKED ISSUE #2796: Inconsistent behavior between `prompter:next-source` and `prompter:previous-source`
`prompter:next-source` returns NIL when the current source is the last one, while `prompter:previous-source` returns the first one when the current one is the first.

`prompter:previous-source` should return NIL when the current source is the first one.

### LINKED ISSUE #2797: Command `last-suggestion` incorrectly handles empty sources
Recipe:

- Issue `set-url`;
- Issue `last-suggestion`;
- Notice that the current suggestion is set to an empty source.

### PR DESCRIPTION
**Pull request by aadcg**
# Description

Bug fixes and addition of prompt buffer commands.

Fixes #2797 
Fixes #2796
Fixes #2742 

TODO:
- [x] Add tests related to how empty sources are handled.

# Discussion

- Keybindings for commands `first-suggestion-within-source` and `last-suggestion-within-source`.
- Are `{first,last}-suggestion-within-source` reasonable command names?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [x] I have pulled from master before submitting this PR
- [x] There are no merge conflicts.
- [x] I've added the new dependencies as:
  - [x] ASDF dependencies,
  - [x] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [x] and Guix dependencies.
- [x] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [x] I have performed a self-review of my own code.
- [x] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [x] Documentation:
  - [x] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [x] I have updated the existing documentation to match my changes.
  - [x] I have commented my code in hard-to-understand areas.
  - [x] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [x] I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - [x] (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [x] Compilation and tests:
  - [x] My changes generate no new warnings.
  - [x] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: source/mode/prompt-buffer.lisp [hunk 1]
```diff
@@ -194,6 +194,22 @@ listed and chosen from with the command `set-action-on-return' (bound to
   (prompter:last-suggestion prompt-buffer)
   (prompt-render-suggestions prompt-buffer))
 
+(define-command-prompt first-suggestion-within-source (prompt-buffer)
+  "Select first entry in the current PROMPT-BUFFER's source."
+  (if (eq (prompter:current-source prompt-buffer)
```
Comment (aartaka) [inline source/mode/prompt-buffer.lisp:199]: Well this condition is quite cryptic. Maybe add a predicate like `last-source-p` and `first-source-p` or check positions of those instead?

Comment (aadcg) [reply source/mode/prompt-buffer.lisp:199]: Positions can't be checked since there may be empty sources. 

I can let-bind the condition.

Comment (aartaka) [reply source/mode/prompt-buffer.lisp]: Yes, that's much better, thank you（＠◇＠）

**Comment by aadcg** [comment]
Ready for review.

CC @Ambrevar @aartaka.

File: libraries/prompter/prompter.lisp [hunk 2]
```diff
@@ -290,9 +293,8 @@ first (resp. last) one of destination source."
 (defun first-suggestion (prompter)
   "Set `current-suggestion' to PROMPTER's first suggestion.
 Empty sources are skipped."
-  (let ((first-non-empty-source
-          (or (find-if #'nonempty-source-p (sources prompter))
-              (first (sources prompter)))))
+  (let ((first-non-empty-source (first (remove-if #'empty-source-p
+                                                  (sources prompter)))))
```
[the remark this answers is not in the document]
Comment (aadcg) [reply libraries/prompter/prompter.lisp:297]: Good point @Ambrevar. I'll add it.

[the remark this answers is not in the document]
Comment (aadcg) [reply libraries/prompter/prompter.lisp:297]: See commits 85dd07a05 and f978a2121.

Is that what you had in mind @Ambrevar?

[the remark this answers is not in the document]
Comment (Ambrevar) [reply libraries/prompter/prompter.lisp:297]: Yup, thanks!

**Comment by aadcg** [comment]
@ambrevar, ok to merge?

**Comment by Ambrevar** [comment]
>     * Keybindings for commands `first-suggestion-within-source` and `last-suggestion-within-source`.

Maybe "shift" the corresponding bindings for `first-suggestion` and `last-suggestion`. 

>     * Are `{first,last}-suggestion-within-source` reasonable command names?

Yes!

OK to merge then!

**Comment by aadcg** [comment]
> Maybe "shift" the corresponding bindings for `first-suggestion` and `last-suggestion`.

Good idea. But I'm not sure what to do with emacs keybindings. Notice that you already need `shift` to hit `M->` or `M-<`.  Any suggestion?

**Comment by Ambrevar** [comment]
Then `M-,` and `M-.` for Emacs?

**Comment by aadcg** [comment]
Those have a key special meaning in Emacs, but I don't have a better suggestion right now so let's stick to that! Anyway, we can always revise it.

On another topic, what about `vi`?

```log
       "g g" 'first-suggestion
       "?" 'first-suggestion-within-source
       "G" 'last-suggestion
       "?" 'last-suggestion-within-source
```

**Comment by aadcg** [comment]
I've found bugs on this PR. Will fix them and add more tests. Not ready to be merged!

**Comment by aadcg** [comment]
Thanks.

@Ambrevar I didn't add VI-specific commands, but feel free to add them.  The ones set for the default keyscheme are available when `vi-mode` is enabled.

**Comment by Ambrevar** [comment]
I haven't used VI-style bindings in so long, at this point I have no clue :p

**Comment by Ambrevar** [comment]
Thanks for this PR, @aadcg !
